inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_b_fs_yx_fsv16_imad.cl

```diff
@@ -0,0 +1,191 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "include/include_all.cl"
+#include "include/data_types.cl"
+
+#define ALIGN_TO(val, multiple) (((val) + (multiple) - 1) / (multiple) * (multiple))
+
+#define AS_TYPE(type, val) CAT(as_, type)(val)
+#define IN_VEC16 MAKE_VECTOR_TYPE(INPUT0_TYPE, 16)
+#define OUT_VEC16 MAKE_VECTOR_TYPE(OUTPUT_TYPE, 16)
+#define CONVERT_OUT CAT(convert_, OUTPUT_TYPE)
+#define CONVERT_OUT_VEC16 CAT(convert_, OUT_VEC16)
+#define FEATURE_SLICE_SIZE 16
+
+#if MAX_POOLING
+    #define INIT_VAL CHAR_MIN
+#elif AVG_POOLING
+    #define INIT_VAL 0
+#else
+#error
+#endif
+
+
+inline int FUNC(apply_pooling)(int tmp, int in)
+{
+#if MAX_POOLING
+    return max(tmp, in);
+#elif AVG_POOLING
+    return tmp + in;
+#endif
+}
+
+KERNEL(pooling_gpu_b_fs_yx_fsv16)(
+    const __global INPUT0_TYPE* input,
+    __global OUTPUT_TYPE* output
+#if HAS_FUSED_OPS_DECLS
+    , FUSED_OPS_DECLS
+#endif
+)
+{
+    const uint x    = (uint)get_global_id(0);
+    const uint y    = (uint)get_global_id(1);
+    const uint bf   = (uint)get_global_id(2);
+    const uint f    = (bf * FEATURE_SLICE_SIZE) % ALIGN_TO(INPUT0_FEATURE_NUM, FEATURE_SLICE_SIZE);
+    const uint b    = (bf * FEATURE_SLICE_SIZE) / ALIGN_TO(INPUT0_FEATURE_NUM, FEATURE_SLICE_SIZE);
+
+    const int offset_x = (int)x*STRIDE_SIZE_X - PADDING_SIZE_X;
+    const int offset_y = (int)y*STRIDE_SIZE_Y - PADDING_SIZE_Y;
+
+    int result[FEATURE_SLICE_SIZE] = { INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL, INIT_VAL };
+
+#ifdef CHECK_BOUNDRY
+    if (offset_x + POOL_SIZE_X < 0 || offset_x >= INPUT0_SIZE_X ||
+        offset_y + POOL_SIZE_Y < 0 || offset_y >= INPUT0_SIZE_Y)
+    {
+        return;
+    }
+
+#ifdef DYNAMIC_KERNEL_DIVIDER
+    uint num_elements = 0;
+#endif
+
+    const uint batch_and_feature_offset = INPUT0_GET_INDEX(b, f, 0, 0);
+    for(uint j = 0; j < POOL_SIZE_Y; j++)
+    {
+        int input_offset_y = offset_y + j;
+        bool zero_y = input_offset_y >= INPUT0_SIZE_Y || input_offset_y < 0;
+        if(!zero_y)
+        {
+            for(uint i = 0; i < POOL_SIZE_X; i++)
+            {
+                int input_offset_x = offset_x + i;
+                bool zero = input_offset_x >= INPUT0_SIZE_X || input_offset_x < 0;
+                if(!zero)
+                {
+                    const uint input_idx = batch_and_feature_offset + input_offset_y*IN_Y_PITCH + input_offset_x*IN_X_PITCH;
+
+                    int4 int_data = vload4(0, (__global int*)(input + input_idx));
```

Comment:
I see that you use vload4 / vstore4 functions for reading and writing data. Did you try to use other ways for this? I don't sure but maybe it'll help to increase performance of your kernel.

Reply:
I just based this kernel on pooling b_fs_yx_fsv4 and bs_fs_yx_bsv16_fsv16 and that is why i used those functions believing it would work well.

Reply:
As agreed above this kernel will not use sub-groups and such experiments can be done later